### PR TITLE
sql/parser: add []byte overloads for hash functions

### PR DESCRIPTION
**Pull request by dt**

### CODE REVIEW

File: pkg/sql/testdata/logic_test/builtin_function [hunk 1]
```diff
@@ -1534,22 +1534,25 @@ six-one-seven
...
-true false true false true false
+true false true false true false true true true
 
-# The hash functions should be stable, so verify that the  following hashes
+# The hash functions should be stable, so verify that the following hashes
 # don't change.
 query T
 SELECT i FROM (VALUES
-  (sha512(NULL)),
+  (sha512(NULL::string)),
```
Comment (dt) [inline pkg/sql/testdata/logic_test/builtin_function]: We could avoid this via a `preferredOverload` hack, but it doesn't seem worth it -- IIRC @knz and @nvanbenschoten would rather avoid using `preferredOverload` except when absolutely forced to for pg compat.

Comment (knz) [inline pkg/sql/testdata/logic_test/builtin_function]: This looks fine to me.

**Review by knz — APPROVED** [review]

**Comment by danhhz** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 1973 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15871#-KjsjhTNKCGMdZKxkHTd:-KjsjhTNKCGMdZKxkHTe:b-fw99c8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e45833c121910f7deaf690059ea777616dff8127/pkg/sql/parser/builtins.go#L1973)):*
> ```Go
> 		var buf string
> 		if d, ok := datum.(*DBytes); ok {
> 			buf = string(*d)
> ```

This causes an unnecessary allocation, no? buf could easily be `[]byte`

---

*[pkg/sql/testdata/logic_test/builtin_function, line 1555 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15871#-KjsDq6go81S2aVtxWaY-r1-1555:-Kjsk3NV2uqLt8mevw8g:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e45833c121910f7deaf690059ea777616dff8127/pkg/sql/testdata/logic_test/builtin_function#L1555)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This looks fine to me.
</blockquote></details>

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15871#-:-Kjsk6_Y6VcQFy5jkPP_:bnfp4nl)*

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 1973 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15871#-KjsjhTNKCGMdZKxkHTd:-Kjsr6qB-WCdU11HqdqI:b31web6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e45833c121910f7deaf690059ea777616dff8127/pkg/sql/parser/builtins.go#L1973)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This causes an unnecessary allocation, no? buf could easily be `[]byte`
</blockquote></details>

It'll be forced to alloc wherever you cast the string to `[]byte` -- doesn't really matter if you do it here or below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15871)*

**Pull request merged**